### ISSUE TITLE
Ability to use other transcript sources for Annotation 

### ISSUE THREAD

**Issue by damiansm** [exomiser/Exomiser#162]
Want to be able to use other sourcs (e.g. Ensembl or Ensembl-gencode transcripts) for Jannovar but Exomiser currently reliant on entrezGene IDs

**Comment by visze** [comment]
I will provide you (tomorrow) with some information about jannovar 0.18 using entrez-gene id

**Comment by damiansm** [comment]
Do we have an ETA for 0.18 Max? The ability to use non-UCSC sources is
becoming quite urgent on my end so trying to plan best way forward

Cheers
Damian

**Comment by visze** [comment]
I think until the mid of december it will be possible to do the release. I have to discuss this with Manuel because he made the most changes and some of them are not stable yet. But he is right now on vacation but will return soon.

You can also make a "hack" by building an ensemble or refseq database and putting the entrez-gene id into the id field of the gene (I did this once and it worked!). The problematic position in exomiser is [here](https://github.com/exomiser/Exomiser/blob/f56cd0ad0dede62432ae1c40d20bbc7f8941ebd7/exomiser-core/src/main/java/de/charite/compbio/exomiser/core/factories/VariantFactory.java#L328-L330) I guess

**Comment by damiansm** [comment]
I will try the hack next week then. Do you want to hold off making the
exomiser changes jules?

**Comment by julesjacobsen** [comment]
Yep, I'll wait on the next Jannovar release so I can do it in one go.

**Comment by visze** [comment]
Manuel and I did a lot of documentation an bug fixing today. So I think we will release Jannovar-0.18 this week.

There are some API changes that will affect exomiser. But I Jules can maybe figure that out. Otherwise I can start a new branch this week try to merge jannovar-0.18-SNAPSHOT and Jules can finish it if it is ready. One API change is that the inheritance stuff is moved into jannovar-core (mendel source folder) and the bridge is in jannovar-htsjdk (mendel source folder). The old is still available but deprecated and we will remove it in the 0.19.

**Comment by julesjacobsen** [comment]
Let me know when you've released it and I'll have a go at migrating over to 0.18. I'll get in touch if I can't figure anything out.

**Comment by visze** [comment]
0.18 is released! It will take up to 24 hours until it will appear on maven central!

**Comment by julesjacobsen** [comment]
Awesome. Looking forward to the improvements.

**Comment by visze** [comment]
Hey. I created a pull request so that you see what is needed for a change. The problem is, that now the tests fail because you cannot use the `geneID` field of a transcript. You have to use the map `altGeneID` with the key `ENTRETZ_ID`

**Comment by julesjacobsen** [comment]
Thanks for the pull request Max - I haven't merged it as I want to fully understand the changes you made and migrate the changes bit by bit. I'm using your PR as a template though so thank you very much for this, it's incredibly useful.

Having made changes to the InheritanceModeAnalyser some of the tests failed, I think due to a bug i the way jannovar handles genotypes from multi-sample VCF. I've opened an [issue](https://github.com/charite/jannovar/issues/285) to address this.

**Comment by visze** [comment]
plz use jannovar 0.19 short bugfix about the entrez-id from ucsc.

Yes I thought this pull request is nice to have a template to see what changed.

I saw the issue about the inhertance stuff. the old inheritance is still in. you can use that. I'm not quite sure if we can fix this in the next weeks and it is very very very unlikely to have a de-novo like that.

**Comment by julesjacobsen** [comment]
OK, but can you merge my PR - this is sufficient to fix one other failure. The other two are long-standing issues which have been present since version 0.16 at least. I agree, maybe vary rare cases, but we are dealing explicitly with very rare cases here and multi-sample VCFs are quite common!

**Comment by damiansm** [comment]
Getting a bit confused here. For running Exomiser using Ensembl
(gencode-only) transcripts should I attempt the hack Max suggested a few
days ago or hold off for 0.19 integration and Jules' new geneIdentifier
class to be able to do the job properly. I probably don't really need this
till after Xmas anyway

**Comment by julesjacobsen** [comment]
Depends on what you really need - if you don't really need it until after Christmas, then I'd wait until this is finished. I don't think that hacking the jannovar database will work. Internally Exomiser is still going to have to use entrez as the id of choice due to the underlying data in the database. Externally it shouldn't make a difference in the end.

What do you want to see in terms of input and output? Define genes using ENSG... identifiers with HGNC symbols and have ENSG, ENST returned in all the output files? Are you going to want to be able to explicitly state what datasource to use e.g ensemble or ucsc as is done with Jannovar? This would be really helpful to have well-defined in advance of making the changes.

I'll get on to the jannovar gene identifiers next week - the inheritance changes have taken me a lot of time to track down the cause of the failing exomiser tests and help fix. I can't say how long it'll take as it affects pretty much the entire system.

@visze Can Jannovar build an ensemble-gencode only or is this the default?

@damiansm If you can define your most important requirement e.g. HTML output must display and link to ENST for transcripts or Gene symbols should allow input using ENSG or HGNC identifiers. I can prioritise that functionality and get you something which gets you sorted, but isn't necessarily full-featured.

**Comment by damiansm** [comment]
Will definitely wait. I just want to run a normal Exomiser/Genomiser
analysis with with an ensembl.ser instead of ucsc.ser so that all
annotations and outputs involve ENSG and ENST IDs rather than the current
UCSC genes and transcripts.

Suspect I will have to download the right subset of gencode-based Ensembl
genes and transcripts and build the .ser file myself but it should have
ENSG and ENST IDs.

**Comment by visze** [comment]
@julesjacobsen jannovar has no default library (you can even build your own transcript database). For hg19 and 'hg20' there are 3 databases available: ucsc, refseq, and ensemble. It would be nice if the user can change the transcript database they need. E.g. we here at the Charité are working with refseq.

For more information howto create the serial file of predefined databases see the [jannovar manual](http://jannovar.readthedocs.io/en/master/download.html)

**Comment by julesjacobsen** [comment]
@visze Yep, I've got this working on a local build. You can specify what jannovar file to run it from so it will use any coordinate set you give it. However it requires that the HGNC file is also incorporated otherwise exomiser will produce nonsense.

Here's the new version using... 
UCSC (``exomiser.transcript-data-file-name=hg19_ucsc.ser``):
```
#CHROM	POS	REF	ALT	QUAL	FILTER	GENOTYPE	COVERAGE	FUNCTIONAL_CLASS	HGVS	EXOMISER_GENE	CADD(>0.483)	POLYPHEN(>0.956|>0.446)	MUTATIONTASTER(>0.94)	SIFT(<0.06)	REMM	DBSNP_ID	MAX_FREQUENCY	DBSNP_FREQUENCY	EVS_EA_FREQUENCY	EVS_AA_FREQUENCY	EXAC_AFR_FREQ	EXAC_AMR_FREQ	EXAC_EAS_FREQ	EXAC_FIN_FREQ	EXAC_NFE_FREQ	EXAC_SAS_FREQ	EXAC_OTH_FREQ	EXOMISER_VARIANT_SCORE	EXOMISER_GENE_PHENO_SCORE	EXOMISER_GENE_VARIANT_SCORE	EXOMISER_GENE_COMBINED_SCORE
10	123256215	T	G	100.0	PASS	0/1	0	missense_variant	FGFR2:uc001lfg.4:c.518A>C:p.(Glu173Ala)	FGFR2	.	0.998	1.0	0.01	.	rs121918506	0.0	.	.	.	.	.	.	.	.	.	.	1.0	0.8762905	1.0	0.9933356
9	94486381	G	A	429.46	PASS	0/1	40	missense_variant	ROR2:uc004arj.2:c.2395C>T:p.(Pro799Ser)	ROR2	.	0.996	1.0	0.094	.	rs141235720	0.442478	0.03994	0.3372	0.0908	0.049329	0.112457	.	0.060551	0.251794	0.37064	0.442478	0.78935075	0.84000266	0.78935075	0.9365846
4	126373317	C	T	1317.59	PASS	0/1	97	missense_variant	FAT4:uc003ifi.1:c.3580C>T:p.(Arg1194Cys)	FAT4	.	0.034	1.0	0.002	.	.	0.110132	.	.	.	.	0.008637	.	.	.	.	0.110132	0.84891415	0.70399815	0.84891415	0.86126685
17	17697093	CCAG	C	398.05	PASS	0/1	33	disruptive_inframe_deletion	RAI1:uc002grm.3:c.870_872del:p.(Gln291del)	RAI1	.	.	.	.	.	rs113303801	0.0	.	.	.	.	.	.	.	.	.	.	0.85	0.64490205	0.85	0.7723361
17	17697404	C	T	730.68	PASS	0/1	52	missense_variant	RAI1:uc002grm.3:c.1142C>T:p.(Ala381Val)	RAI1	.	0.696	0.99992	0.002	.	rs113208290	0.542123	0.2995	0.4651	0.1589	0.134875	0.406153	.	0.060478	0.542123	0.060569	0.443459	0.7672179	0.64490205	0.85	0.7723361
16	2287649	A	C	145.37	PASS	0/1	13	missense_variant	DNASE1L2:uc002cpn.3:c.527A>C:p.(Asp176Ala)	DNASE1L2	.	0.831	.	0.029	.	rs62621282	0.393125	0.2596	0.2908	0.0761	0.074313	0.393125	.	0.146951	0.364289	0.007651	0.135135	0.77630925	0.7036929	0.77630925	0.76058507
```
ENSEMBL (``exomiser.transcript-data-file-name=hg19_ensembl.ser``);
```
#CHROM	POS	REF	ALT	QUAL	FILTER	GENOTYPE	COVERAGE	FUNCTIONAL_CLASS	HGVS	EXOMISER_GENE	CADD(>0.483)	POLYPHEN(>0.956|>0.446)	MUTATIONTASTER(>0.94)	SIFT(<0.06)	REMM	DBSNP_ID	MAX_FREQUENCY	DBSNP_FREQUENCY	EVS_EA_FREQUENCY	EVS_AA_FREQUENCY	EXAC_AFR_FREQ	EXAC_AMR_FREQ	EXAC_EAS_FREQ	EXAC_FIN_FREQ	EXAC_NFE_FREQ	EXAC_SAS_FREQ	EXAC_OTH_FREQ	EXOMISER_VARIANT_SCORE	EXOMISER_GENE_PHENO_SCORE	EXOMISER_GENE_VARIANT_SCORE	EXOMISER_GENE_COMBINED_SCORE
10	123256215	T	G	100.0	PASS	0/1	0	missense_variant	FGFR2:ENST00000336553:c.1421A>C:p.(Glu474Ala)	FGFR2	.	0.998	1.0	0.01	.	rs121918506	0.0	.	.	.	.	.	.	.	.	.	.	1.0	0.8762905	1.0	0.9933356
9	94486381	G	A	429.46	PASS	0/1	40	missense_variant	ROR2:ENST00000375708:c.2395C>T:p.(Pro799Ser)	ROR2	.	0.996	1.0	0.094	.	rs141235720	0.442478	0.03994	0.3372	0.0908	0.049329	0.112457	.	0.060551	0.251794	0.37064	0.442478	0.78935075	0.84000266	0.78935075	0.9365846
4	126373317	C	T	1317.59	PASS	0/1	97	missense_variant	FAT4:ENST00000335110:c.6040C>T:p.(Arg2014Cys)	FAT4	.	0.034	1.0	0.002	.	.	0.110132	.	.	.	.	0.008637	.	.	.	.	0.110132	0.84891415	0.70399815	0.84891415	0.86126685
17	17697093	CCAG	C	398.05	PASS	0/1	33	disruptive_inframe_deletion	RAI1:ENST00000261641:c.870_872del:p.(Gln291del)	RAI1	.	.	.	.	.	rs113303801	0.0	.	.	.	.	.	.	.	.	.	.	0.85	0.64490205	0.85	0.7723361
17	17697404	C	T	730.68	PASS	0/1	52	missense_variant	RAI1:ENST00000261641:c.1142C>T:p.(Ala381Val)	RAI1	.	0.696	0.99992	0.002	.	rs113208290	0.542123	0.2995	0.4651	0.1589	0.134875	0.406153	.	0.060478	0.542123	0.060569	0.443459	0.7672179	0.64490205	0.85	0.7723361
```
REFSEQ (``exomiser.transcript-data-file-name=hg19_refseq.ser``):
```
```

**Comment by visze** [comment]
looks good to me!

**Comment by julesjacobsen** [comment]
Cheers @visze, that was a really helpful new feature in Jannovar. Seems to work really well too. Fist bump to you and @holtgrewe!
